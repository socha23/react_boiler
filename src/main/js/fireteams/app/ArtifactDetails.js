import React from 'react'
import {connect} from 'react-redux'

import {findArtifactByTagId} from '../../artifacts/selectors'

const ArtifactDetails = ({artifact}) => <div>
    {artifact.name}
</div>;


export default connect((state, {fireteam}) => ({
    artifact: findArtifactByTagId(state, fireteam.targetTagId)
}))(ArtifactDetails);

