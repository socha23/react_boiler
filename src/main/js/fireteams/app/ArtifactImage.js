import React from 'react'

import contextPath from '../../common/contextPath'

const ArtifactImage = ({artifact}) => artifact.images ?
    <div style={{flex: "1 1 auto", overflow: "hidden"}}>
        <img style={{maxWidth: "100%", maxHeight: "100%"}}
            src={contextPath() + "/api/images/" + artifact.images[0].id}/>
    </div>
    : <div/>;

export default ArtifactImage


