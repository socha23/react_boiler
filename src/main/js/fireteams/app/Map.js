import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'

import {getFireteamFloorPlan, getFireteamTag, getTargetTag} from '../selectors'

import {DotMarker} from '../../maps/Marker'
import TransformMatrix from '../../common/components/TransformMatrix'
import * as transform from '../../common/transformMatrixHelpers'
import {TypeArtifact, TypeFireteam} from "../../tags/TagType";

import translatePoint from '../../common/translatePoint'


export class MapComponent extends React.Component {

    static propTypes = {
        fireteamTag: PropTypes.object.isRequired,
        targetTag: PropTypes.object,
        floorPlan: PropTypes.object,
        zoom: PropTypes.number,
        onTargetShown: PropTypes.func,
    };

    static defaultProps = {
        floorPlan: {},
        targetTag: {},
        zoom: 1,
        onTargetShown: () => {}
    };

    state = {
        imgBounds: null
    };

    componentWillReceiveProps = (nextProps) => {
        if (!nextProps.targetTag) {
            nextProps.onTargetShown(false);
        } else {
            const targetScreenPos = this.posToScreen(nextProps.targetTag.position);
            nextProps.onTargetShown(targetScreenPos.y > 0);
        }
    };

    setImgBounds = (e) => {
        this.setState({
            imgBounds: {
                topLeft: {x: 0, y: 0},
                // clientWidth i clientHeight podają wymiary przed transformacją, a o to nam tu chodzi
                bottomRight: {x: e.target.clientWidth, y: e.target.clientHeight}
            }
        });
    };

    posToScreen = (point) => {
        let fireteamOnImg = this.toImgCoords(this.props.fireteamTag.position);
        let pointOnImg = this.toImgCoords(point);

        const matrix = transform.multiply(
                    transform.translateMatrix(-fireteamOnImg.x, -fireteamOnImg.y),
                    transform.scaleMatrix(this.props.zoom),
                    transform.rotateMatrix(this.getRotation()),
                    transform.translateMatrix(this.getFocalPoint().x, this.getFocalPoint().y),
                );

        return transform.transformPoint(matrix, pointOnImg);
    };


    toImgCoords = (point) => {
        if (this.state.imgBounds) {
            return translatePoint(
                this.props.floorPlan,
                this.state.imgBounds,
                point
            );
        } else {
            return point;
        }
    };

    getFocalPoint = () => {
        return this.containerElem ? {
            x: this.containerElem.clientWidth / 2,
            y: this.containerElem.clientHeight - 100
        } : {
            x: 0, y: 0
        };
    };

    getRotation = () => {
        if (this.props.fireteamTag && this.props.targetTag) {
            const fireteamPos = this.toImgCoords(this.props.fireteamTag.position);
            const targetPos = this.toImgCoords(this.props.targetTag.position);
            return Math.atan2(fireteamPos.x - targetPos.x, fireteamPos.y - targetPos.y);
        } else {
            return 0;
        }
    };

    renderTargetMarker = () => {
        if (!this.props.targetTag) {
            return <span/>;
        }
        const targetPos = this.posToScreen(this.props.targetTag.position);
        return <DotMarker x={targetPos.x} y={targetPos.y} color={TypeArtifact.color}/>
    };

    render = () => {

        const fireteamTranslation = this.toImgCoords(this.props.fireteamTag.position);
        const fireteamPos = this.posToScreen(this.props.fireteamTag.position);

        return <div
                ref={e => {this.containerElem = e}}
                style={{
                    height: "100%",
                    position: "relative",
                    top: 0,
                    left: 0,
                    zIndex: 0,
                    overflow: "hidden",
                    backgroundColor: "white",
                    ...this.props.style

                }}
            >
                <TransformMatrix
                    translateX={this.getFocalPoint().x}
                    translateY={this.getFocalPoint().y}
                >
                        <TransformMatrix
                            scale={this.props.zoom}
                            rotation={this.getRotation()}
                            translateX = {-fireteamTranslation.x}
                            translateY = {-fireteamTranslation.y}
                        >
                            <img
                                onLoad={this.setImgBounds}
                                src={this.props.floorPlan.base64content}/>
                        </TransformMatrix>
                </TransformMatrix>
            {this.renderTargetMarker()}
            <DotMarker x={fireteamPos.x} y={fireteamPos.y} color={TypeFireteam.color}/>
        </div>
    }
}

export default connect((state, {fireteam}) => ({
    fireteamTag: getFireteamTag(state, fireteam),
    targetTag: getTargetTag(state, fireteam),
    floorPlan: getFireteamFloorPlan(state, fireteam)
}))(MapComponent)