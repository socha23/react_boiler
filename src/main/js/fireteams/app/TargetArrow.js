import React from 'react'
import {connect} from 'react-redux'

import {getFireteamTag, getTargetTag} from '../selectors'

import DistanceBetweenTags from '../../tags/DistanceBetweenTags'

export const TargetArrow = ({fireteamTag, targetTag, style={}}) => {
    if (!targetTag) {
        return <span/>
    }

    return <div style={{position: "absolute", width: "100%", top: style.top || 0, left: 0, zIndex: 2, textAlign: "center"}}>
        <div className="borderBottomPulseBlue" style={{
            borderBottomWidth: 80,
            borderBottomStyle: "solid",
            borderLeft: "100px solid transparent",
            borderRight: "100px solid transparent",
            width: 0,
            height: 0,
            margin: "auto",
            zIndex: 1
        }}/>
        <div className="pulseBlue" style={{width: 120, height: 40, margin: "auto", zIndex: 1}}/>
        <div style={{position: "absolute", top: 40, zIndex: 2, width: "100%", color: "white", fontSize: 60}}>
            <DistanceBetweenTags from={fireteamTag} to={targetTag} showUnit={false}/>
        </div>
    </div>;

    //if (targetTag.areaName == fireteamTag.areaName) {
};


export default connect((state, {fireteam}) => ({
    fireteamTag: getFireteamTag(state, fireteam),
    targetTag: getTargetTag(state, fireteam)
}))(TargetArrow)

