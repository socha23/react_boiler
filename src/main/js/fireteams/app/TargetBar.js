import React from 'react'
import {connect} from 'react-redux'

import {getTargetTag} from '../selectors'

export const TargetBarComponent = ({targetTag}) => <div style={{
    backgroundColor: "black",
    padding: 10,
    color: "white",
    fontSize: 30

}}>
    <div style={{
        display: "flex",
        justifyContent: "center",
        alignItems: "center"
    }}>
        {targetTag ? targetTag.label : "Brak wyznaczonego celu"}
    </div>
</div>;


export default connect((state, {fireteam}) => ({
    targetTag: getTargetTag(state, fireteam)
}))(TargetBarComponent)

