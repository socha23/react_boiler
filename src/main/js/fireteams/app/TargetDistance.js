import React from 'react'
import {connect} from 'react-redux'

import {getFireteamTag, getTargetTag} from '../selectors'

import DistanceBetweenTags from '../../tags/DistanceBetweenTags'

export const DistanceComponent = ({fireteamTag, targetTag}) => {
    if (targetTag.areaName == fireteamTag.areaName) {
        return <SameFloorDistance fireteamTag={fireteamTag} targetTag={targetTag}/>

    } else {
        return <DifferentFloorDistance fireteamTag={fireteamTag} targetTag={targetTag}/>
    }
};


export default connect((state, {fireteam}) => ({
    fireteamTag: getFireteamTag(state, fireteam),
    targetTag: getTargetTag(state, fireteam)
}))(DistanceComponent)


const SameFloorDistance = ({fireteamTag, targetTag}) => <span style={{
    backgroundColor: "black",
    padding: 10,
    color: "white",
    fontSize: 50,
    position: "absolute",
    top: 0,
    left: 0,
    zIndex: 1
}}>
    <DistanceBetweenTags from={fireteamTag} to={targetTag}/>
</span>;

const DifferentFloorDistance = ({fireteamTag, targetTag}) => <div style={{
    backgroundColor: "black",
    padding: 10,
    color: "white",
    fontSize: 30

}}>
    <DistanceBetweenTags from={fireteamTag} to={targetTag}/>
    <span style={{marginLeft: 10}}>
        ({targetTag.areaName})
    </span>
</div>;
