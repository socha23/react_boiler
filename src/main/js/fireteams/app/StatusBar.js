import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'

import Fullscreen from '../../common/components/Fullscreen'

import {getFireteamTag, getTargetTag} from '../selectors'

import Map from './Map'



export const StatusBarComponent = ({fireteamTag, targetTag, targetName}) => <BlackBar>
    {
        targetTag ?
            <Target fireteamTag={fireteamTag} targetTag={targetTag}/>
            :
            <NoTarget/>
    }
</BlackBar>;


export default connect((state, {fireteam}) => ({
    fireteamTag: getFireteamTag(state, fireteam),
    targetTag: getTargetTag(state, fireteam)
}))(StatusBarComponent)

const BlackBar = ({children}) => <div style={{
        backgroundColor: "black",
        paddingTop: 10,
        color: "white",
        fontSize: 30

}}>
    <div style={{
        display: "flex",
        justifyContent: "center",
        alignItems: "center"
    }}>
        {children}
    </div>
</div>;

const NoTarget = () => <div>
    Brak wyznaczonego celu
</div>;

const Target = ({fireteamTag, targetTag}) => <div>
    {targetTag.label}
</div>;

