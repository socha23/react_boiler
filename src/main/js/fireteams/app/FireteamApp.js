import React from 'react'
import {connect} from 'react-redux'


import Fullscreen from '../../common/components/Fullscreen'

import Map from './Map'
import TargetBar from './TargetBar'
import TargetDistance from './TargetDistance'
import ArtifactDetails from './ArtifactDetails'
import {distance} from '../../tags/DistanceBetweenTags'

import {getFireteamTag, getTargetTag} from '../selectors'


const MapMode = ({fireteam}) => {
    return <Fullscreen style={{display: "flex", flexDirection: "column"}}>
        <Map fireteam={fireteam}/>
        <TargetDistance fireteam={fireteam}/>
        <TargetBar fireteam={fireteam}/>
    </Fullscreen>
};

const ArtifactMode = ({fireteam}) => {
    return <Fullscreen style={{display: "flex", flexDirection: "column"}}>
        <TargetBar fireteam={fireteam}/>
        <div style={{flexGrow: 1}}>
            <ArtifactDetails fireteam={fireteam}/>
        </div>
    </Fullscreen>
};



function switchToArtifactMode(fireteamTag, targetTag, maxDistance) {
    return targetTag.type == "artifact"
        && fireteamTag.areaName == targetTag.areaName
        && distance(fireteamTag, targetTag) <= maxDistance
}

let ModeSwitcher = ({fireteam, fireteamTag, targetTag, distanceForArtifactMode}) =>
    switchToArtifactMode(fireteamTag, targetTag, distanceForArtifactMode)
        ? <ArtifactMode fireteam={fireteam}/>
        : <MapMode fireteam={fireteam}/>;

ModeSwitcher = connect((state, {fireteam}) => ({
    targetTag: getTargetTag(state, fireteam),
    fireteamTag: getFireteamTag(state, fireteam)
}))(ModeSwitcher);

const FireteamApp = ({fireteam}) => <ModeSwitcher fireteam={fireteam} distanceForArtifactMode={5}/>;


export default FireteamApp;
