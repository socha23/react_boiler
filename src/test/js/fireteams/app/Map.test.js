import React from 'react'
import {testSnapshot} from '../../testUtils'

import {MapComponent} from 'fireteams/app/Map'

it('renders correctly', testSnapshot(
            <MapComponent
                fireteam={{id: "ft1", name: "Team Alpha", tagId: "tA", targetTagId: "tB"}}
                fireteamTag={{id: "tA", position: {x: 100, y: 100}, coordinateSystemId: "coordId"}}
                targetTag={{id: "tB", position: {x: 200, y: 200}, coordinateSystemId: "coordId"}}
                floorPlan={{id: "coordId", base64content: "content", topLeft: {x: 0, y: 0}, bottomRight: {x: 400, y: 400}}}
            />
));
