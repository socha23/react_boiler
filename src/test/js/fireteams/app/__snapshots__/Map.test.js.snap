// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<MapComponent
  fireteam={
    Object {
      "id": "ft1",
      "name": "Team Alpha",
      "tagId": "tA",
      "targetTagId": "tB",
    }
  }
  fireteamTag={
    Object {
      "coordinateSystemId": "coordId",
      "id": "tA",
      "position": Object {
        "x": 100,
        "y": 100,
      },
    }
  }
  floorPlan={
    Object {
      "base64content": "content",
      "bottomRight": Object {
        "x": 400,
        "y": 400,
      },
      "id": "coordId",
      "topLeft": Object {
        "x": 0,
        "y": 0,
      },
    }
  }
  targetTag={
    Object {
      "coordinateSystemId": "coordId",
      "id": "tB",
      "position": Object {
        "x": 200,
        "y": 200,
      },
    }
  }
>
  <div
    style={
      Object {
        "backgroundColor": "white",
        "height": "100%",
        "left": 0,
        "overflow": "hidden",
        "position": "relative",
        "top": 0,
        "zIndex": 0,
      }
    }
  >
    <TransformMatrix
      originX={0}
      originY={0}
      rotation={0}
      scale={1}
      translateX={0}
      translateY={0}
    >
      <div
        style={
          Object {
            "transform": "matrix(1,0,0,1,0,0)",
            "transformOrigin": "top left",
          }
        }
      >
        <TransformMatrix
          originX={0}
          originY={0}
          rotation={-2.356194490192345}
          scale={1}
          translateX={0}
          translateY={0}
        >
          <div
            style={
              Object {
                "transform": "matrix(-0.7071067811865475,-0.7071067811865476,0.7071067811865476,-0.7071067811865475,0,0)",
                "transformOrigin": "top left",
              }
            }
          >
            <TransformMatrix
              originX={0}
              originY={0}
              rotation={0}
              scale={1}
              translateX={-100}
              translateY={-100}
            >
              <div
                style={
                  Object {
                    "transform": "matrix(1,0,0,1,-100,-100)",
                    "transformOrigin": "top left",
                  }
                }
              >
                <img
                  onLoad={[Function]}
                  src="content"
                />
              </div>
            </TransformMatrix>
            <DotMarker
              color="red"
              x={0}
              y={0}
            >
              <Marker
                color="red"
                x={0}
                y={0}
              >
                <div
                  style={
                    Object {
                      "fontSize": 14,
                      "left": 0,
                      "lineHeight": 1.4,
                      "position": "absolute",
                      "top": 0,
                      "zIndex": 1,
                    }
                  }
                >
                  <div
                    style={
                      Object {
                        "backgroundColor": "red",
                        "border": "1px solid black",
                        "borderRadius": 15,
                        "cursor": "pointer",
                        "height": 30,
                        "left": -15,
                        "position": "relative",
                        "top": -15,
                        "width": 30,
                      }
                    }
                  />
                  <span />
                </div>
              </Marker>
            </DotMarker>
            <DotMarker
              color="blue"
              x={100}
              y={100}
            >
              <Marker
                color="blue"
                x={100}
                y={100}
              >
                <div
                  style={
                    Object {
                      "fontSize": 14,
                      "left": 100,
                      "lineHeight": 1.4,
                      "position": "absolute",
                      "top": 100,
                      "zIndex": 1,
                    }
                  }
                >
                  <div
                    style={
                      Object {
                        "backgroundColor": "blue",
                        "border": "1px solid black",
                        "borderRadius": 15,
                        "cursor": "pointer",
                        "height": 30,
                        "left": -15,
                        "position": "relative",
                        "top": -15,
                        "width": 30,
                      }
                    }
                  />
                  <span />
                </div>
              </Marker>
            </DotMarker>
          </div>
        </TransformMatrix>
      </div>
    </TransformMatrix>
  </div>
</MapComponent>
`;
